experimental/iterators/include/iterators/Dialect/Iterators/IR/IteratorsOps.td

```diff
@@ -292,6 +292,30 @@ def Iterators_ReduceOp : Iterators_Op<"reduce",
   }];
 }
 
+def Iterators_ScanColumnarBatchOp : Iterators_Op<"scan_columnar_batch",
```

Comment:
This seems dangerous as it seems to go into an N^2 design, I also see a FullTableScanOp and a ColumnarScanOp.

Why not just a single ScanOp that adapts to the type of its argument?

Reply:
What are the two dimensions? For each memory format, we need a way to iterator over the elements (i.e., "scan" it); that's only one dimension.

I don't have a strong opinion about overloading argument types, yet, but my initial reflex is to have different ops for different input types. I expect them to require relatively different lowerings, so a single op would essentially consists of a large switch and then different logic. (For similar enough formats, where the lowering is essentially the same, or once we have an abstraction that makes them the same, my arguments don't apply, so I'd agree with having a single op.)

Reply:
`that's only one dimension` yes that's precisely why I am suggesting a ScanOp: we shouldn't add a second dimension that is "just the type": a scan can be polymorphic.

As to the big switch aspect: experience shows that this is a symptom of a missing interface.
Once we start seeing swtiches grow, it makes sense to extract out the interfaces we want to see.
If you already know what the interface looks like even better but that is generally hard to foresee.

Reply:
I see that this is an interesting part of the design space. I don't see things clearly enough, yet, though, to foresee what possible design make sense. Can we defer the discussion until we introduce scans for other formats?

Reply:
Notes from an offline meeting with @nicolasvasilache:

* There is a terminology issue with the word `scan`. In functional programming, a scan is a generalized prefix sum; in databases, a scan is the operator that reads the data from a base table. We thus need a better term and document the difference in terminology.
* It is still an open question whether we should have one polymorphic "read base table op" or a dedicated one per data format. The lowering for each format may be different, and there may be different optimizations that apply depending on the format, but maybe some or all of that can be done with interfaces.

Reply:
Connecting this to discussions we currently have in a different PR:
For columnar formats, there is also the dimension of does the scan access all columns of a given table or only parts. So essentially, should we hand the pointers to all the columns and tell the operation which ones to read from, or should we only hand it the ones that it should actually read from.

Reply:
Some thoughts about naming:

* A DB "scan" operator takes a single value, for example a columnar batch, and helps us *iterate* over them. So names like `iterate_columnar_batch` could be appropriate. Python has functions as `iter_values`, `iter_keys`, and `iter_items` on its dictionaries, for example, and I think they serve the same purpose.
* In a similar spirit, these "scan" operators turn a single value into an `iterator.stream`, so `columnar_batch_to_stream` could express that.
* On a different note, there may be different ways to iterate over the same thing: for a columnar batch, one might want to iterate over slices of the batch (`iter_slices`?) in order to do something similar to tiling, or over individual tuples (`iter_tuples`?). In a tensor or memref, one might want to do the same thing: iterate over smaller tensors/memrefs (in different orders) or over individual numbers.